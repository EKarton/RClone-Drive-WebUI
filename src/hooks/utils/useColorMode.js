import { useContext } from 'react';
import { ColorModeContext } from 'contexts/ColorMode/index';

export default function useColorMode() {
  const context = useContext(ColorModeContext);

  if (context === undefined) {
    throw new Error('useColorMode() needs to be used inside ColorModeProvider');
  }

  return context;
}
