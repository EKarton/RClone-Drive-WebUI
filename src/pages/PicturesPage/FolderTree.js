import TreeView from "@mui/lab/TreeView";
import TreeItem from "@mui/lab/TreeItem";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import ChevronRightIcon from "@mui/icons-material/ChevronRight";
import useRCloneClient from "hooks/useRCloneClient";
import { useEffect, useState } from "react";
import { StatusTypes } from "pages/PicturesListPage";

const FolderTree = ({ remotes, onFolderSelect }) => {
  const handleNodeSelect = (_e, nodeIds) => {
    onFolderSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} />
      ))}
    </TreeView>
  );
};

const FolderTreeItem = ({ remote, curPath, label }) => {
  const rCloneClient = useRCloneClient();

  const [status, setStatus] = useState(StatusTypes.LOADING);
  const [subFolders, setSubFolders] = useState(undefined);
  const [error, setError] = useState(null);

  const [isExpanded, setIsExpanded] = useState(false);

  useEffect(() => {
    const fetchFolders = async () => {
      try {
        setError(null);
        setStatus(StatusTypes.LOADING);

        const data = await rCloneClient.fetchSubFolders(remote, curPath);

        setSubFolders(data.sort());
        setStatus(StatusTypes.SUCCESS);
      } catch (err) {
        setError(err);
      }
    };

    if (isExpanded) {
      fetchFolders();
    }
  }, [curPath, isExpanded, rCloneClient, remote]);

  const handleTreeItemClicked = () => {
    setIsExpanded(true);
  };

  const renderSubFolders = () => {
    if (!isExpanded) {
      return <div />;
    }

    if (status === StatusTypes.LOADING) {
      return <div>Loading</div>;
    }

    if (status === StatusTypes.ERROR) {
      return <div>Error! {error}</div>;
    }

    if (subFolders.length === 0) {
      return <div>No subfolders</div>;
    }

    return subFolders.map((subFolder) => (
      <FolderTreeItem
        remote={remote}
        curPath={subFolder.Path}
        label={subFolder.Name}
      />
    ));
  };

  return (
    <TreeItem
      nodeId={`${remote}:${curPath}`}
      label={label}
      onClick={handleTreeItemClicked}
      expandIcon={<ChevronRightIcon />}
      collapseIcon={<ExpandMoreIcon />}
    >
      {renderSubFolders()}
    </TreeItem>
  );
};

export default FolderTree;
