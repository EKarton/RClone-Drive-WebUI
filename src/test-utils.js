import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
