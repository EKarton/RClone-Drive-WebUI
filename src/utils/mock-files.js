export const mockFiles = {
  list: [
    {
      Path: '.rclone',
      Name: '.rclone',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-11-01T21:09:05.237Z',
      IsDir: true,
      ID: '1a4mI4vsGKypqFFPwLnrFid1ZdvlhirDi',
    },
    {
      Path: 'Documents',
      Name: 'Documents',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T01:51:05.982Z',
      IsDir: true,
      ID: '1_za5Kqy9ZjXZ7rsztGpKhtJLTuNCKv8q',
    },
    {
      Path: 'Games',
      Name: 'Games',
      Size: -9,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T05:14:59.237Z',
      IsDir: true,
      ID: '1j6mAjxQoJokiU8yWJYgyWYIiSIgH1G7M',
    },
    {
      Path: 'Literature',
      Name: 'Literature',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-29T07:10:05.059Z',
      IsDir: true,
      ID: '1dj1T5YkjRQZNiQ5efCA8PzideJb88d3K',
    },
    {
      Path: 'Pictures',
      Name: 'Pictures',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:39:11.850Z',
      IsDir: true,
      ID: '1D2PpmoIT23NqyTnolmohZ0UxvOd2D7n4',
    },
    {
      Path: 'Software',
      Name: 'Software',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:45:02.471Z',
      IsDir: true,
      ID: '1zzr0Bwp9tel66rCUKz8mZIXRBR2kuywr',
    },
    {
      Path: 'Videos',
      Name: 'Videos',
      Size: -8,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T04:22:24.997Z',
      IsDir: true,
      ID: '1eJlJ1PT7b9XY3Q5tLeU49BNkug259986',
    },
    {
      Path: 'backup.sh',
      Name: 'backup.sh',
      Size: 1110,
      MimeType: 'application/x-sh',
      ModTime: '2021-11-14T18:58:44.655Z',
      IsDir: false,
    },
  ],
};

export const mockRemotes = {
  remotes: ['googledrive', 'onedrive', 'icloud'],
};
