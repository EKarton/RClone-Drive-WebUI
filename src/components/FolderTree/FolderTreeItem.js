import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import TreeItem from '@mui/lab/TreeItem';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { StatusTypes } from 'utils/constants';

export default function FolderTreeItem({ remote, curPath, label }) {
  const rCloneClient = useRCloneClient();

  const [status, setStatus] = useState(null);
  const [subFolders, setSubFolders] = useState();

  const [isExpanded, setIsExpanded] = useState(false);

  useEffect(() => {
    const fetchFolders = async () => {
      try {
        setStatus(StatusTypes.LOADING);

        const data = await rCloneClient.fetchSubFolders(remote, curPath);

        setSubFolders(data.sort());
        setStatus(StatusTypes.SUCCESS);
      } catch (err) {
        setStatus(StatusTypes.ERROR);
      }
    };

    if (isExpanded) {
      fetchFolders();
    }
  }, [curPath, isExpanded, rCloneClient, remote]);

  const handleTreeItemClicked = () => {
    setIsExpanded(true);
  };

  const renderSubFolders = () => {
    if (!isExpanded || !status) {
      return <div />;
    }

    if (status === StatusTypes.LOADING) {
      return <div>Loading</div>;
    }

    if (status === StatusTypes.ERROR) {
      return <div>Error!</div>;
    }

    if (subFolders.length === 0) {
      return <div>No subfolders</div>;
    }

    return subFolders.map((subFolder) => (
      <FolderTreeItem
        remote={remote}
        curPath={subFolder.Path}
        label={subFolder.Name}
        key={subFolder.Name}
      />
    ));
  };

  return (
    <TreeItem
      nodeId={`${remote}:${curPath}`}
      label={label}
      onClick={handleTreeItemClicked}
      expandIcon={<ChevronRightIcon />}
      collapseIcon={<ExpandMoreIcon />}
    >
      {renderSubFolders()}
    </TreeItem>
  );
}

FolderTreeItem.propTypes = {
  remote: PropTypes.string,
  curPath: PropTypes.string,
  label: PropTypes.string,
};
