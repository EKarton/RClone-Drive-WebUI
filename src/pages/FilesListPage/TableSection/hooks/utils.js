import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
