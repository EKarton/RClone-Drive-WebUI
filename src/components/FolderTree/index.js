import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import TreeView from '@mui/lab/TreeView';
import PropTypes from 'prop-types';
import FolderTreeItem from './FolderTreeItem';

export default function FolderTree({ remotes, onSelect }) {
  const handleNodeSelect = (_e, nodeIds) => {
    onSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} key={remote} />
      ))}
    </TreeView>
  );
}

FolderTreeItem.propTypes = {
  remotes: PropTypes.arrayOf(PropTypes.string),
  onSelect: PropTypes.func,
};
