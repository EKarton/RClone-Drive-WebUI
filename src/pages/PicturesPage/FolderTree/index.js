import TreeView from '@mui/lab/TreeView';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import FolderTreeItem from './FolderTreeItem';

export default function FolderTree({ remotes, onFolderSelect }) {
  const handleNodeSelect = (_e, nodeIds) => {
    onFolderSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} />
      ))}
    </TreeView>
  );
}
