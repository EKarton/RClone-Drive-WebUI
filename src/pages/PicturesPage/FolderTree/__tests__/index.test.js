import FolderTreeItem from '../FolderTreeItem';
import FolderTree from '..';
import { customRender, fireEvent, screen } from 'test-utils';
import { mockRemotes } from 'utils/mock-files';
import { TreeItem } from '@mui/lab';

jest.mock('../FolderTreeItem');

describe('FolderTree', () => {
  beforeEach(() => {
    FolderTreeItem.mockImplementation(({ remote }) => {
      return <TreeItem nodeId={remote} label={remote} data-testid={remote} />;
    });
  });

  it('should match snapshot given list of remotes', () => {
    const { baseElement } = customRender(
      <FolderTree remotes={mockRemotes.remotes} onFolderSelect={jest.fn()} />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should call onFolderSelect() given user clicks on a FolderTreeItem', async () => {
    const onFolderSelect = jest.fn();
    customRender(
      <FolderTree remotes={mockRemotes.remotes} onFolderSelect={onFolderSelect} />
    );

    fireEvent.click(screen.getByText(mockRemotes.remotes[0]));

    expect(onFolderSelect).toBeCalledWith(mockRemotes.remotes[0]);
  });
});
