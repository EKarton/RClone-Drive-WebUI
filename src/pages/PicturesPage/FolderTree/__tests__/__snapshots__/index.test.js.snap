// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderTree should match snapshot given list of remotes 1`] = `
<body>
  <div>
    <ul
      aria-label="file tree navigator"
      aria-multiselectable="false"
      class="MuiTreeView-root css-161j9ok-MuiTreeView-root"
      role="tree"
      tabindex="0"
    >
      <li
        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
        data-testid="googledrive"
        role="treeitem"
        tabindex="-1"
      >
        <div
          class="css-1g86id8-MuiTreeItem-content MuiTreeItem-content"
        >
          <div
            class="MuiTreeItem-iconContainer"
          />
          <div
            class="MuiTreeItem-label"
          >
            googledrive
          </div>
        </div>
      </li>
      <li
        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
        data-testid="onedrive"
        role="treeitem"
        tabindex="-1"
      >
        <div
          class="css-1g86id8-MuiTreeItem-content MuiTreeItem-content"
        >
          <div
            class="MuiTreeItem-iconContainer"
          />
          <div
            class="MuiTreeItem-label"
          >
            onedrive
          </div>
        </div>
      </li>
      <li
        class="MuiTreeItem-root css-16alkdk-MuiTreeItem-root"
        data-testid="icloud"
        role="treeitem"
        tabindex="-1"
      >
        <div
          class="css-1g86id8-MuiTreeItem-content MuiTreeItem-content"
        >
          <div
            class="MuiTreeItem-iconContainer"
          />
          <div
            class="MuiTreeItem-label"
          >
            icloud
          </div>
        </div>
      </li>
    </ul>
  </div>
</body>
`;
